presto-hive/src/main/java/io/prestosql/plugin/hive/BackgroundHiveSplitLoader.java

```diff
@@ -314,37 +316,44 @@ private void invokeNoMoreSplitsIfNecessary()
                 throw new PrestoException(NOT_SUPPORTED, "Bucketed table in SymlinkTextInputFormat is not yet supported");
             }
 
-            // TODO: This should use an iterator like the HiveFileIterator
-            ListenableFuture<?> lastResult = COMPLETED_FUTURE;
-            for (Path targetPath : getTargetPathsFromSymlink(fs, path)) {
-                // The input should be in TextInputFormat.
-                TextInputFormat targetInputFormat = new TextInputFormat();
-                // the splits must be generated using the file system for the target path
-                // get the configuration for the target path -- it may be a different hdfs instance
-                FileSystem targetFilesystem = hdfsEnvironment.getFileSystem(hdfsContext, targetPath);
-                JobConf targetJob = toJobConf(targetFilesystem.getConf());
-                targetJob.setInputFormat(TextInputFormat.class);
-                targetInputFormat.configure(targetJob);
-                FileInputFormat.setInputPaths(targetJob, targetPath);
-                InputSplit[] targetSplits = targetInputFormat.getSplits(targetJob, 0);
-
-                InternalHiveSplitFactory splitFactory = new InternalHiveSplitFactory(
-                        targetFilesystem,
-                        partitionName,
-                        inputFormat,
-                        schema,
-                        partitionKeys,
-                        effectivePredicate,
-                        partition.getColumnCoercions(),
-                        Optional.empty(),
-                        isForceLocalScheduling(session),
-                        s3SelectPushdownEnabled);
-                lastResult = addSplitsToSource(targetSplits, splitFactory);
-                if (stopped) {
-                    return COMPLETED_FUTURE;
+            return hdfsEnvironment.doAs(hdfsContext.getIdentity().getUser(), () -> {
```

Comment:
Wouldn't it be sufficient to wrap `targetInputFormat.getSplits` only?

Reply:
@findepi I tried this but product-tests-suite-6-non-generic fails with an Auth error when calling `getTargetPathsFromSymlink()`.  Now I've changed the code to wrap the two calls individually instead of the entire block, let me know what you think.